app/export.go

```diff
@@ -63,82 +64,143 @@ var exportablePreferences = map[imports.ComparablePreference]string{{
 	if err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting channels")
-	if err = a.exportAllChannels(writer, teamNames); err != nil {
+	if err = a.exportAllChannels(zipWriter, teamNames); err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting users")
-	if err = a.exportAllUsers(writer); err != nil {
+	if err = a.exportAllUsers(zipWriter); err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting posts")
-	attachments, err := a.exportAllPosts(ctx, writer, opts.IncludeAttachments)
+	attachments, err := a.exportAllPosts(ctx, zipWriter, opts.IncludeAttachments)
 	if err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting emoji")
-	emojiPaths, err := a.exportCustomEmoji(writer, outPath, "exported_emoji", !opts.CreateArchive)
+	emojiList, err := a.exportCustomEmoji(zipWriter, outPath, "exported_emoji", !opts.CreateArchive)
 	if err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting direct channels")
-	if err = a.exportAllDirectChannels(writer); err != nil {
+	if err = a.exportAllDirectChannels(zipWriter); err != nil {
 		return err
 	}
 
 	ctx.Logger().Info("Bulk export: exporting direct posts")
-	directAttachments, err := a.exportAllDirectPosts(ctx, writer, opts.IncludeAttachments)
+	directAttachments, err := a.exportAllDirectPosts(ctx, zipWriter, opts.IncludeAttachments)
 	if err != nil {
 		return err
 	}
-
+	var missingAttachments []*imports.LogFormat
+	var logData *imports.LogFormat
 	if opts.IncludeAttachments {
-		ctx.Logger().Info("Bulk export: exporting file attachments")
+		mlog.Info("Bulk export: exporting file attachments")
 		for _, attachment := range attachments {
-			if err := a.exportFile(outPath, *attachment.Path, zipWr); err != nil {
+			if logData, err = a.ExportAttachments(attachment, outPath, zipWr); err != nil {
 				return err
 			}
+			if logData != nil {
+				missingAttachments = append(missingAttachments, logData)
+			}
 		}
 		for _, attachment := range directAttachments {
-			if err := a.exportFile(outPath, *attachment.Path, zipWr); err != nil {
+			if logData, err = a.ExportAttachments(attachment, outPath, zipWr); err != nil {
 				return err
 			}
+			if logData != nil {
+				missingAttachments = append(missingAttachments, logData)
+			}
 		}
-		for _, emojiPath := range emojiPaths {
-			if err := a.exportFile(outPath, emojiPath, zipWr); err != nil {
+		for _, emoji := range emojiList {
+			if logData, err = a.ExportAttachments(emoji, outPath, zipWr); err != nil {
 				return err
 			}
+			if logData != nil {
+				missingAttachments = append(missingAttachments, logData)
+			}
 		}
 	}
 
+	if logWriter != nil && len(missingAttachments) != 0 {
+		fileData, err := json.MarshalIndent(imports.MissingAttachments{MissingAttachments: missingAttachments}, "", " ")
+		if err != nil {
+			return model.NewAppError("exportLogFile", "Failed to marshal and indent the data", nil, err.Error(), http.StatusInternalServerError)
```

Comment:
We generally use error id and translates here. You can check other occurrences in the code.